### DISCUSSION TITLE
Using u-blox ZED F9-P for full raw data collection via Pygpsclient

### DISCUSSION THREAD

**Discussion by savioratharv** [semuconsulting/PyGPSClient#147] [Q&A] [answered]
can I use a u-blox ZED F9-P module and obtain .ubx files using Pygpsclient? I need it to be a .ubx file containing full raw data, so that I can use this data for PPK and achieve cm level accuracy.
I am very new to this stuff so any help is appreciated.
Thanks in advance.

**Comment by semuadmin** [comment, marked as answer]
Hi @savioratharv 

Yes, you can certainly do that. Just select the 'DataLogging' option in PyGPSClient with a format of 'binary' _before_ you open the serial connection. You will be prompted to select the folder into which the log file will be saved. PyGPSClient will save the raw data from the receiver (NMEA, UBX or RTCM) to a timestamped binary file in that folder e.g. `pygpsdata-20240816085025.log`. If needs be, you can change the file suffix to `*.ubx`, but this doesn't affect the content at all.

<img width="287" alt="Screenshot 2024-08-16 at 08 50 56" src="https://github.com/user-attachments/assets/9051f735-a657-4010-9453-93bf90a605c3">

You can then read back the contents of the file using the FILE input mode of PyGPSClient.

<img width="58" alt="Screenshot 2024-08-16 at 08 58 46" src="https://github.com/user-attachments/assets/81a14f82-aaf8-4ff4-adbb-0e122e971dff">

FYI you can also do the same thing with the [gnssdump](https://github.com/semuconsulting/pygnssutils?tab=readme-ov-file#gnssdump) command line utility if you'd prefer not to use a GUI application.

NB: note that the F9P receiver only outputs NMEA data by default. If you specifically need UBX data, you will need to configure the receiver accordingly. PyGPSClient has preset facilties to suppress NMEA data and enable a minimal cohort of UBX messages instead:

<img width="392" alt="Screenshot 2024-08-16 at 09 04 08" src="https://github.com/user-attachments/assets/4a492076-ad5c-47c7-babd-ca4e4ed82c95">

**Reply by savioratharv** [reply]
Thank you so much for replying! Can I load this txt file of configurations as I need all the data that would be necessary for PPK. https://github.com/u-blox/ublox-C099_F9P-uCS/blob/master/zed-f9p/F9P%20Base%20config%20C99.txt

**Reply by semuadmin** [reply]
> Thank you so much for replying! Can I load this txt file of configurations as I need all the data that would be necessary for PPK. https://github.com/u-blox/ublox-C099_F9P-uCS/blob/master/zed-f9p/F9P%20Base%20config%20C99.txt

PyGPSClient does support the loading of u-center *.txt files via its CFG Configuration Load/Save/Record facility:

<img width="361" alt="Screenshot 2024-08-16 at 09 57 59" src="https://github.com/user-attachments/assets/ddba7f7b-04af-4c91-9aa5-a690f6a3750e">

**_However_**, the specific file you've attached appears to be an older unsupported version from 2019. If you need a current F9P base station configuration in *txt format which is supported by PyGPSClient, try one of these (**check the file is compatible with your receiver's firmware version e.g. HPG 1.32**):

https://www.ardusimple.com/configuration-files/

e.g. https://www.ardusimple.com/wp-content/uploads/2022/09/simpleRTK2B_FW132_Base-00.txt

**Reply by savioratharv** [reply]
Thank you so much!

**Reply by savioratharv** [reply]
Also, what all options should I choose here for my purpose?
![image](https://github.com/user-attachments/assets/4613ca55-950b-4b9b-9a7c-e030126861fc)

**Reply by semuadmin** [reply]
> Also, what all options should I choose here for my purpose? ![image](https://private-user-images.githubusercontent.com/91827941/358534217-4613ca55-950b-4b9b-9a7c-e030126861fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4MDAyOTEsIm5iZiI6MTcyMzc5OTk5MSwicGF0aCI6Ii85MTgyNzk0MS8zNTg1MzQyMTctNDYxM2NhNTUtOTUwYi00YjliLTlhN2MtZTAzMDEyNjg2MWZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE2VDA5MTk1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIyZDI5MDUzNmQxY2I5OTZmYjVkNDUzNTQ4OTQ4YjEzY2YzNjEwN2RmNzlhMjI4ZDNiMGQ2MjQwNzMzMDMwYmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.BfS_puUqcEOnECD0f4CFqxyGqXZY1zQvSIhYqrvJAp8)

I don't know - what is your purpose? 🤔😉

**Reply by savioratharv** [reply]
lol 😂
I wish to collect all relevant data such that I would be able to perform PPK using the ubx file generated. Is the "Protocols Shown" option only applicable for messages shown in the console display? Or does it also control the data being logged?

**Reply by semuadmin** [reply]
> lol 😂 I wish to collect all relevant data such that I would be able to perform PPK using the ubx file generated. Is the "Protocols Shown" option only applicable for messages shown in the console display? Or does it also control the data being logged?

OK so I'm not sure how familiar you are with PPK workflows, so apologies if I'm stating the obvious here, but much of the functionality in PyGPSClient is geared towards RTK (this is, real-time kinematics, as opposed to post-processing kinematics) using real-time correction data from NTRIP or SPARTN sources. I am by no means an expert in PPK and I'm not sure to what extent PyGPSClient would help you in this context.

The majority of PPK workflows and tooling are predicated on obtaining raw navigation and observation data from GNSS base and/or rover receivers in RINEX format, along with other fixed reference information. Unfortunately, PyGPSClient doesn't natively support RINEX, but it does support the UBX RXM-SFRBX and UBX RAW-RAWX message types which contain raw navigation and observational data in ubx (binary) format. There are a number of tools available to convert this data to RINEX format (e.g. RTKLIB), but I can't really offer any assistance with those.

In answer to your question above, the Protocols Shown options only affect what appears in the console (_so you can suppress data you're not interested in_) - they don't affect what the receiver is actually outputting.

I recommend you familiarise yourself with the F9P Interface Specification to understand the content of the various UBX NAV and RXM message types.

https://www.u-blox.com/sites/default/files/documents/u-blox-F9-HPG-1.50_InterfaceDescription_UBXDOC-963802114-12815.pdf

For general queries on F9P functionality and applications, I would refer you to the u-blox community forum:

https://portal.u-blox.com/s/topic/0TO2p000000Hr7dGAC/positioning

**Reply by savioratharv** [reply]
Thanks a lot for the info! I will look into the specific message types more. I would be using either RTKLIB or Emlid studio for PPK and I am getting correction data from the correction stations in my country. The module is in transit and will be delivered in a few days. I will reach out if I face any issues. I really appreciate the help!

**Reply by savioratharv** [reply]
Hello, the module arrived and we have tested it. Pygpsclient works brilliantly and is an absolute godsend for our project. We were just facing an issue that I wanted to ask. The GPS module is itself capable of providing sub-meter level accuracy in our testing but when we do PPK on the raw data using correction data from CORS (the national GPS correction network in my country), the accuracy doesn't seem to improve and it is still sub-meter level accuracy and not cm level accuracy. The CORS station we are using for corrections is 25.8km away from our testing area, could that be the reason? Or is there something else that might be causing the issue? Also when collecting data using Pygpsclient, the dashboard shows horizontal and vertical GPS accuracy, how is this accuracy calculated?

**Reply by semuadmin** [reply]
> Hello, the module arrived and we have tested it. Pygpsclient works brilliantly and is an absolute godsend for our project. We were just facing an issue that I wanted to ask. The GPS module is itself capable of providing sub-meter level accuracy in our testing but when we do PPK on the raw data using correction data from CORS (the national GPS correction network in my country), the accuracy doesn't seem to improve and it is still sub-meter level accuracy and not cm level accuracy. The CORS station we are using for corrections is 25.8km away from our testing area, could that be the reason? Or is there something else that might be causing the issue? Also when collecting data using Pygpsclient, the dashboard shows horizontal and vertical GPS accuracy, how is this accuracy calculated?

There are any number of reasons why you may not be achieving cm-level accuracy - without knowing the specifics of your configuration, I can only offer general advice - see for example this recent [Discussion thread](https://github.com/semuconsulting/PyGPSClient/discussions/152) and [RTK-TIPS.md](https://github.com/semuconsulting/PyGPSClient/blob/master/RTK-TIPS.md).

The horizontal and vertical accuracy figures are calculated within the u-blox receiver's firmware and output via e.g. the UBX NAV-PVT message as `hAcc` and `vAcc` data attributes - they are not calculated by PyGPSClient. AFAIK the details of this calculation are proprietary, but my understanding is that it is essentially based on a statistical analysis of:
- Number of satellites
- Signal strengths
- Geometric distribution of the satellites (DOP values)
- Pseudorange residuals ([UERE](https://en.wikipedia.org/wiki/Error_analysis_for_the_Global_Positioning_System))

There have been various discussions on this topic on the [u-blox community forum](https://portal.u-blox.com/s/question/0D52p0000CYAG4WCQX/how-vertical-and-horizontal-accuracies-are-calculated-), to which I would refer you. I should warn you - this is an extremely complex topic! There is a high-level wiki [here](https://www.semuconsulting.com/gnsswiki/#Errors).

**Reply by gdt** [reply]
Also, you say "accuracy is not cm".   Be careful not to confuse actual accuracy with error estimates.   Assessing actual accuracy requires a monument that you have an accurate position for.     However, another technique I find useful is to repeatedly measure a monument, on different days, and to plot all the measurements, and see show they cluster.  There could still be bias, but it is very enlightening.

You said PPK and CORS (do you mean US National CORS Network?), but you didn't explain how you are doing PPK, and what kind of residuals and float/fix indication the PPK solution is giving you.   I have done RTK with the F9P's processing, and  not PPK, and would expect better performance with PPK compared to RTK since it can propagate fixed integer ambiguitiies backwards in time rather than just forward.

**Reply by savioratharv** [reply]
I am using the Indian CORS network. As you said about measuring accuracy, we have a reference station within our institution that is not operational but whose exact GPS coordinates are known. I am planning to test out the GPS accuracy properly using that. But as it stands, the results I am getting are not able to trace the path taken by me correctly and when I tested both the native GPS coordinates and the obtained coordinates after PPK, I noticed the native GPS coordinates were more representative of the path than the points after PPK. I tried using both Emlid Studio and RTKLIB for PPK. RTKLIB tells the accuracy of the points after PPK is sub-metre.
[query.zip](https://github.com/user-attachments/files/16957142/query.zip)
I am attaching a zip file with both the raw data in ubx format and the correction data from CORS. The area where we tested this has a lot of buildings and a canopy of trees, could that also be a reason for this result? And the fact that the base station we are using for corrections is 25.8km away.

**Reply by semuadmin** [reply]
Hi @savioratharv,

So, having taken a quick look at your raw ubx data, I can see straight away that your satellite coverage is unlikely to be sufficient for cm accuracy via RTK. Your carrier-to-noise (CNo) ratios are far too low and your sky coverage appears to be concentrated in a roughly 36° vertical arc.

You're seeing this:

<img width="624" alt="Screenshot 2024-09-11 at 08 37 19" src="https://github.com/user-attachments/assets/74e34f1d-e52c-49db-b998-9c306a6b6d27">

... but you should be seeing something more like this, with much broader sky visibility and around 15-20 satellites with CNo > 45 dB:

<img width="624" alt="Screenshot 2024-09-11 at 08 36 33" src="https://github.com/user-attachments/assets/1a799962-e907-4773-b709-32b604424e01">

At a push I'd say you need to elevate your antenna and/or use a more powerful active antenna with a decent LNA. Suggest having another look through [RTK-TIPS](https://github.com/semuconsulting/PyGPSClient/blob/master/RTK-TIPS.md).

In terms of any PPK results via RTKLIB, you need to be asking those questions in the [u-blox](https://portal.u-blox.com/s/) or [FOSS GPS](https://lists.osgeo.org/listinfo/foss-gps) forums rather than here.

**Reply by gdt** [reply]
Also, you should understand what datum your national CORS operates in.   I have no idea, but I would expect that India has a modern (after the dawn of the satellite era) datum that is plate fixed.   which is not WGS84.

**Reply by semuadmin** [reply]
I'm going to lock this discussion at this point, before we go off on too many tangents.

@savioratharv if you have any further queries in relation to PyGPSClient, don't hesitate to come back to the Discussion Forum.
